### ISSUE TITLE
MAX - Proper Projects

### ISSUE THREAD

**Issue by OldFlak** [TheGameCreators/GameGuruRepo#4294]
Several builds back the splash said we now have Projects 
- we do have most of the required folders in each project within projectbank
- but putting media in there does nothing. The media doesn't show up in Object Library.

So is this idea canned or just broken atm?

Just need to know so I can continue copying everything over for my game 
- so are project files meant to work
- or should we just use the 'user' folder in each section

-----------------

Would really appreciate Proper Projects

That would solve all the current issues and the ones inherited from classic when creating standalones 
ie:
- copying media over that the game does not use
- and not copying over media it depends on

Projects are more than just good house keeping 
- it also makes it easy to ensure your game has everything it needs for standalone
- everything the game uses is in the projects folder hierarchy
- everything in there gets copied to standalone

No idea if it is really feasible for the majority of users tho :)

**Comment by MonkeyFrogStudio** [comment]
I have to admit, I don't think I understand how these project folders are supposed to work. Are we supposed to manually dump stuff there ourselves? If so, then that's not really a project folder or how it's supposed to work.

I created a test level to send to Preben so he could suss out a bug. MAX created a project folder in projectbank named after my project. Great! I added an object to the level, saved, and then looked in the project folder. Do you know what's in there under Files? Nothing but empty folders. The object placed in my level was not copied into the projectbank>Files>entitybank folder. The script the object uses was not transferred to the scriptbank folder there. Nothing was.

The idea of a project folder is that everything that pertains to your game is THERE in that ONE PLACE, not scattered between MAX's install folders, your own user folders, and in all the DLC folders from DLCs you've purchased. The idea is that if you add an object to a level in your project, that object gets COPIED into the project folder of that project's name and then gets REFERENCED in the game engine for that project so that it looks for that object THERE, in the project folder, and not somewhere else. This means that we can BACKUP that entire project folder (and nothing else) and, as a result, back up our entire game ... all objects used in that project, all scripts used in that one project, all images, etc. 

We shouldn't have to manually set all this up, manually transfer stuff to the project folder, etc. It should do it for us. Anytime I make a change to my game, that project folder should reflect it - objects added, objects removed, scripts used, etc. Right now, what is it? How does it even work? I don't get it.

**Comment by OldFlak** [comment]
@MonkeyFrogStudio Yes - that's how I expect it to work as well.

But currently it just creates the folder hierarchy when you create your project and will again if you delete them 
- but nothing goes in there
- if you tell MAX to put models in there when you import new ones, it returns an empty path, 
- - still puts the models in there, but they are not accessible via the Object Library
- and anything you put in there manually as you mentioned is not accessible in MAX

My guess is it is either broken, or not implemented atm, despite the splash saying it has been added.

**Comment by MonkeyFrogStudio** [comment]
It is a bit maddening, in a way. When you work with other game engines, they do this automatically. You start a new project in Unreal, it creates a folder for that project and a basic structure there. You can add your own folders, too, but they end up in that one project folder. In the end, you can backup or even delete that ONE project folder and you have backed up everything pertaining to that project (or deleted it, if that's what you wanted to do). 

S2, Flax, Unreal, etc. They all work this way. I have a Temp folder that I use for testing. I know that anything in that folder is "temporary" and it's okay to delete it at any time. Often, when testing something in another game engine, I will create the project in that Temp folder. This way, I know that whatever I do in that game engine for that project, it will all be saved to that one, single project folder there. And when I am done testing, I can simply delete it and be done with it. 

MAX simply needs to work like this, imo.

**Comment by OldFlak** [comment]
Yes - couldn't agree more
I had started putting stuff in the user folders for each area - but to me it is not really ideal
Then we had hopes of the Project folder idea - but no working atm

As a result I have no idea where to place my custom stuff at all now.

So even tho V1 is not far off, can't create any real levels yet, because I don't want to keep shifting media around - too much work in that. And I don't want it all over the place either.

**Comment by davetgc** [comment]
I'm marking this as a bug, as the behaviour isn't as expected. Should the dev team have a clearer explanation, they'll correct the labels and let you all know.,

**Comment by OldFlak** [comment]
@davetgc thanks for the feedback Dave :)

**Comment by LeeBamberTGC** [comment]
@OldFlak Just going through your issue now in sequence:

1. We may have fixed this since your report, but I added an entity into a folder folder inside a project called Milk: 
![image](https://user-images.githubusercontent.com/35840583/232123526-b0f3fa7f-884f-4b82-9aab-1e1791992944.png)
When launching MAX and opening the project, and then clicking Add, and the Project Media, the entity shows:
![image](https://user-images.githubusercontent.com/35840583/232123629-66add036-5c89-40f9-a8ed-b0ed2f276320.png)
Selecting it adds it to the level:
![image](https://user-images.githubusercontent.com/35840583/232123705-05ef4277-3a26-4427-b957-a64250a7e584.png)

2. Noi need to manually create a USER folder, create any folder name you want inside the entitybank, just make sure you do not use the same name as one of the existing out-of-the-box stock folders as your stuff will mix in with the custom stuff when you search for objects in the Object Library:
![image](https://user-images.githubusercontent.com/35840583/232123990-988bbea6-587c-45c8-a2a0-e87f0ca6083c.png)

3. If we copied all the media used in your game project over to the project folder, each project folder would become massive (+GBs). The way it works now, only one copy of the media is stored safely, and when standalones are created, they are copied from that reference. Only new custom media should be stored in the project, again so only one copy of the media exists on your PC and keeping storage as efficient as possible.

@MonkeyFrogStudio Feedback on your post:

1. The project folder is created automatically when you create a new project from the HUB area, no need to manually create this folder. It is also populated with all the project-specific files the game needs, such as storyboard layouts, collection and quest lists, thumbnails for the project, etc.   Most users will never even know there is a project folder and all happens automatically through the MAX UI, managing project settings as they move from project to project.  In this way it works like Unreal, creating the project for you and putting game project-specific files there.  

2. What MAX will NOT do is copy over duplicates of all the media used in every level of that game project, it would become insanely big and the extra copy steps would slow down operations and produce needless redundancy.  Instead MAX has references to all the stock and DLC media, kept safe and protected in the Steam root folder of the product, using them as needed for your many game projects.

3. By default, imported models are placed in the generic writables folder structure, that way your imported models are available to all current and future game projects you may be working on. Forcing imported models into just the current game project is fine for that game, but the user will have no access to it when they make a new project.  This way it stays referencable once they have imported and saved their new object.

4. If you however want to ensure your imported model goes just to your current project, just set the folder path and the object will be saved there, accessible only for that project you are working on: 
![image](https://user-images.githubusercontent.com/35840583/232126863-610a9d63-a1e6-4941-84a8-e58ffa042000.png)
And available in the Object Library from its project-bound location:
![image](https://user-images.githubusercontent.com/35840583/232126943-cb6c848c-17ac-4f20-94f9-b9cfc35923d0.png)

5. Advice on where you place your custom media, I would want to keep it accessible to all my game projects, and so would create a new folder, maybe call it "MyOwnStuff" in the writables entitybank (i.e. "Documents\GameGuruApps\GameGuruMAX\Files\entitybank\MyOwnStuff"). Or if you want it just for a single project, you can create it here ("Documents\GameGuruApps\GameGuruMAX\Files\projectbank\Milk\Files\entitybank\MyOwnStuff"). Once you have your location sorted, just copy all your custom media in there, or use the importer as described above to populate either of those folders.

So that is the way it works now, and how it was designed to work to save storage space, avoid duplications and still provide multiple ways to manage your media (root location for MAX and DLC stock, writables for your own custom stuff and project folder for game specific custom stuff).  The majority of users will never need to enter or even see these folders, it is all handled through the HUB.  For users who are creating their own custom media OUTSIDE of MAX, then we may need an additional section in the user guide.

If you find any bugs and the system does not behave as I have described, let me know here and I can go through and make sure it works as designed.  For now I will set this as an enhancement discussion so we can continue figuring out what works for everybody.  Some ideas:

A. An extra tick box in Import Model that targets the current project folder directly, so no need to find the folder, it just adds it right in there without ever needing to navigate to it.

B. A tick box when creating the project to insist that MAX copies EVERYTHING over that is used, including objects, textures, sounds, etc., and add a warning note to state this feature will consume multiple GBs for large projects.

C. Add a UI that lists the entire tree structure of the project files, so users can manually prune old files they no longer require, and in the mode described in [B], only use project folder files to make standalone exports so the user can control what goes into it, along with a file size reading for each file including some color-coded categories to show where the total storage size comes from (similar to how mobile phones show you how you are using your phone storage).

**Comment by OldFlak** [comment]
Hi Lee and thanks for looking at this. 
Is does indeed work as you describe above but that is not project based - projects only have media used by the specified project.

So projects:
- the folders are replicated in each game project exactly as done in the project\files\folder hierarchy 
- the media in each game project is only accessible by going to the projects folder in the library

- inside the various main folders we can use any name for each sub-folder
- MAX would use projects\projectname to separate all media from every other folder in the project library

Examples
-- when searching entities: projects\naelurec\subfolders for entities
-- when searching audio: projects\naelurec\subfolders for audio

Here be a mock up
![image](https://user-images.githubusercontent.com/12246318/232206614-970a2b2d-89cd-4d66-802f-be06900cd7aa.png)

Point 2 and 3 above: 
- for me I don't plan on using any stock media
- but if I do I wouldn't mind it being in every one of my projects, a have a single drive solely for game dev - lots of space to go
- when MAX is ready to produce shareable games I will only have one or two projects anyway going at a time
- I am more concerned with the plethora of media that I won't use taking up space on my system drive

Point 5 above:
- this is no different to just using the User folders - if we do that then projects are kinda pointless

Points A and B would be very welcome indeed, C would be a nice feature and would be a good thing to add in the future - for now I am happy to house keep under the hood.

Again thanks for looking at this :)

**Comment by MonkeyFrogStudio** [comment]
I agree with OldFlak. If we're going to have project folders at all, then they actually need to work like actual project folders. 

As far as hard drive space - most people, I would think, are not going to be working on more than one actual game at a time. I really don't see people having three or four game projects going on. Therefore, if MAX is ensuring everything that is part of that project is in the project folder, even if it's a copy of a stock asset, then it's only ONE additional copy. Not dozens, for sure. Once the end-user is finished with their game, they can then back up that ONE project folder and have every single asset (model, material, script, audio, video, etc.) backed up with it. If they ever need to access that game again, it's all there in one place. 

Doing this would also keep that one game's files completely separate from MAX's assets. Let's say that you are using a particular script in your game and it's working just as you like it. Then, for some reason, with an update of MAX, that script gets updated and it no longer works as you wanted/needed it. No worries. The "copy" in your project folder is preserved and works as intended. You can move on with your game not having to worry about possible changes that could affect your work.

You could also make changes to the FPE files in your game's project folder without affecting the FPE files that come stock with MAX.

Like OldFlak, I don't plan to use any of the assets that come with MAX, but plan to create everything I want to use for myself. So, MAX, in this instance, is already hogging my hard drive space, filling it with content that I cannot really remove because, if I do, then with the next update it will just be added back in again. The stock MAX entitybank install folder has over 12GB of content in it ... 12GB that I never plan to use in any of my games. So, if you're really worried about the end-user's hard drive space, then why haven't you given as a means to not download all this stuff (if we don't want/need it) and to keep it from ever being installed in the first place?

Concerning point 3 - I would counter that once a project folder is created imported content SHOULD be placed automatically in the project folder's folder structure by default. That's actually what the project folder should be for. What I would suggest is that there be an option (could be an Advanced setting or some such that you tick) that says, "Use Reference instead of Placing Imports in Project" (or whatever best makes sense grammatically). So, by default, all new imports would go into the project folder. But there could also be an option the end-user could tick to make it so imports go to the user area and a reference link is placed in the project folder instead. 

If I need an object or objects from a particular project to use in a new or different project, I could always go into that first project folder and copy over the sub-folder I need to the new project, thus giving me access to another project's items. Or perhaps MAX could have a LINK option. What I mean by a LINK option is there could be a means of LINKing a current project to another project so the end-user could access another project's items. Or a means of copying over assets from one project to another.

Concerning point 4 - If this is what we have to do, then why bother with having a project folder at all? We can already do this manually on our own.

Concerning point 5 - But the point of a project folder is to keep stuff organized for your one, single project, not to make stuff accessible to all projects/games! Even with project folders, this is an option. The end-user can decide that they want a particular asset to be available for all projects. If so, they can override the default of being imported into the project folder and place the imported object somewhere else ... or copy it over for such a purpose.

Points A and B - Both are fine. Options are always good, imo.
Point C - Sounds good, too, actually.

**Comment by 3coms** [comment]
Still need to track every media loaded via scripts (images, sounds, etc)
**LoadImage( "scriptbank\\images\\Morning Mountain Stroll\\001.png" )**

Or I use this one in my convo script to store all the char convos.
**file = assert(io.open("Convo System\\unique_audio_list.txt", "r"))** 

I use this loop to load all my sounds in a level:
       **for k,v in pairs(Audio_load) do
		LoadGlobalSound ( v , k )
	end**
Here "**Audio_load.txt**" contains all the audio files path.

So, all those files should be present in my project, but it is not that easy for the engine to be aware of, at least it tracks every script I'm using in my project.

**Comment by LeeBamberTGC** [comment]
@OldFlak Let's assume we have the option to bin everything currently done related to the user-facing aspects of the Project system as it exists currently (i.e remove the Project Media in the object library, no FIles folder in the projectbank area, etc). Can you describe how your Project System would work, as seen from the end users' perspective? I.e. how they interact with it, what they need to do, how it looks, etc. I do not think we can morph what we have now with the feedback here. Clearly I do not understand what everyone here wants. I think a brand new system is needed that starts with first principals, do you fancy starting the design discussion, and we can go from there?

**Comment by 3coms** [comment]
Still think that duplicating stuff is not an option, but just saving a reference to the files, not the files themselves, so, you don't duplicate stuff and don't no use a lot of hdd space in your pc.
So, if the user wants to make changes to any mesh, he may do in his favorite 3d soft and reimport it, since that mesh belongs to a project (maybe some tickbox in the importer to indicate it) so, GGMax updated the mesh reference it has in its DB.
And when the user decides to make a standalone file, so, GGMax may use those references to be aware of the files it has to copy to a standalone file.
However, still we are going to need to track the stuff loaded via script in any way or give the user the way to add those files manually, maybe by categories, so, it makes life easier to mGGMax, though.

**Comment by OldFlak** [comment]
@LeeBamberTGC Exactly as MonkeyFrog and I explained above - but if it helps we can start fresh - do you want a new thread started for this?

@3coms that is not how projects work - you copy all media to your project and can edit it how ever you want without affecting any stock media - so for example if you made duplicates of existing models and called them something else so you can edit them you still used the same amount of space on you hdd

**Comment by MonkeyFrogStudio** [comment]
I just want to connect this request here, too:

https://github.com/TheGameCreators/GameGuruRepo/issues/4479

It would help free up hard drive space for a lot of people if the end-user were able to download only what they needed. If project folders actually copied over complete assets, then this would work great together. For example, let's say that all the MAX content is a DLC. I could then download the Arctic Collection, for example, use only ONE of the objects from it. It would get copied over to the project folder for that project. I could then remove the Arctic Collection DLC completely, freeing up all that hard drive space, but still have access to that ONE asset that I wanted. As it is now, if I want to use only ONE asset from the Arctic Collection, then I have to have ALL of the MAX content installed ... all 12+GB!

**Comment by 3coms** [comment]
**then I have to have ALL of the MAX content installed ... all 12+GB!**

I'm afraid about it. :)

----------------------------------------------------------
I might be wrong or don't get the point right, but so, this means I'm going to have the same table mesh in let's say 6 different project folders? to say the less, because in my case it is going to be a ton of GBs. lol

**Comment by MonkeyFrogStudio** [comment]
Why on earth would you have six different project folders? If you're even halfway serious about game design, then you'll be working on a single game for more than a year (probably 2 or more, especially if you're by yourself). Work on one game, get it finished, back it up, then work on the next. Simple.

But if you've bothered to actually read this thread, then you'd see that Lee was proposing an OPTION for this, not that it would be mandatory for everyone. So, why object if this feature would not affect you in the slightest?

**Comment by 3coms** [comment]
We can't talk just about us, it is something to be used by many people, at least I think so, and not everybody has your workflow, I can work on 5 different projects and use the same mesh in all of them, and this no means the 5 projects are going to be a serious game.
And I stand here since this post can convert into a chat, and that's not the thread goal.

**Comment by LeeBamberTGC** [comment]
@OldFlak Yes a new issue might be best with yours' and MonkeyFrogStudio's vision of the perfect project management system for MAX.  @3coms I agree that many users will not be needing this system right away, but I do think a good 10% of users will eventually want to create a serious long term game project, and this system should help with that, and the games produced this way should go on to become poster children for what MAX can really do.  Happy to support this one. It won't be rushed into V1, but we have all of 2023 to do it right :)

**Comment by MonkeyFrogStudio** [comment]
> We can't talk just about us, it is something to be used by many people, at least I think so, and not everybody has your workflow, I can work on 5 different projects and use the same mesh in all of them, and this no means the 5 projects are going to be a serious game.
> And I stand here since this post can convert into a chat, and that's not the thread goal.

You do realize that is EXACTLY what I was saying to you - not everyone has YOUR workflow. Lee is offering workflows for you and for me. That was my point. You're making it for me. ;)

**Comment by 3coms** [comment]
I already wrote the script that :
1. locate every script used by the project
2. Track every script by searching for media loaded
3. List all the media (images. audio, etc) loaded via a script, in a plane file. (file paths)
4. Unfortunately LUA does not have copy/paste media (no text, but images, audio, video, etc)
5. Some sort of **SendKey** (**CTRL+C** and **CTRL+V**) or at least "SetGamePlayerStatePlrKeyForceKeystate()" working as expected, we can create some workaround and copy/paste the needed media in its respective project folders, (audio in audiobank, img in imagebank, and so on)
6.

**Comment by LeeBamberTGC** [comment]
@3coms I believe you sent me a separate email on this last point, we can resume conversation there. For the rest of this issue, I will close it as I anticipate a new issue being created that describes a whole new way to manage projects properly, not Lees' crazy upside-down view of it :)

**Comment by 3coms** [comment]
https://www.youtube.com/watch?v=zPH3ovItsXI
